[IOTDB-560] add TSRecordOutputFormat to write TsFile via Flink DataSet/DataStream API.

WeiZhong94 opened the pull request:
TsFile is a columnar storage file format in Apache IoTDB. It is designed for time series data and supports efficient compression and query and is easy to be integrated into big data processing frameworks.

Apache Flink is a framework and distributed processing engine for stateful computations over unbounded and bounded data streams and becoming more and more popular in IOT scenes. So, it would be great to integrate IoTDB and Flink.

This pull request adds a TSRecordOutputFormat to support write TsFiles on flink via DataStream/DataSet API.

More detail can be found in discussion thread [1], or in the IoTDB wiki [2]

[1]https://lists.apache.org/thread.html/r6dd6afe4e8e4ca42e3ddfbc80609597788f90b214e7a81788c3b51b3%40%3Cdev.iotdb.apache.org%3E

[2]https://cwiki.apache.org/confluence/display/IOTDB/%5BImprovement+Proposal%5D+Add+Flink+Connector+Support+for+TsFile

sunjincheng121 commented:
Thanks for the PR @WeiZhong94 !
Sorry, I'm busy with the Flink forward SF meeting. I'll review this PR as soon as possible next week！
BTW：Could you please have look at this PR  @qiaojialin @jixuan1989 :)

sunjincheng121 commented:
@WeiZhong94 Could you please rebase the PR. and I would like to have look at it then.

qiaojialin commented on flink-tsfile-connector/src/main/java/org/apache/iotdb/flink/tsfile/RowTSRecordConverter.java:
```diff
@@ -0,0 +1,157 @@
+			DataPoint templateDataPoint = templateRecord.dataPointList.get(dataPointIndexMapping[i]);
+			Object o = input.getField(i);
+			if (o != null) {
+				Class typeClass = o.getClass();
```
templateDataPoint.type could be used to switch the data type.

WeiZhong94 commented:
@sunjincheng121 @qiaojialin Thanks for your review! Sorry for the late reply because my work is really busy recently :(. I have addressed your comments, please take a look.

> Though the "OutputFormat" can be used in DataStream API, it would be better if we support writing TsFile via "StreamingFileSink", which is integrated with the checkpointing mechanism to provide exactly once semantics. Of course we could do this in follow up PR.

Yes, "StreamingFileSink" is much better than "OutputFormat" for streaming job. To support writing TsFile via "StreamingFileSink", we need to implement the flink interface "org.apache.flink.api.common.serialization.BulkWriter". Current blocker is that the BulkWriter wraps the output target as a "FSDataOutputStream" object. It is possible to write TsFile data via "FSDataOutputStream" but needs further discussion as the "FSDataOutputStream" does not support the "truncate" method, which is required by the "TsFileOutput" interface.

sunjincheng121 commented:
Hi @WeiZhong94 I am fine with current `OutputFormat` solution for now. and please have look at the CI issue. Thanks.

qiaojialin commented on flink-tsfile-connector/README.md:
```diff
@@ -91,3 +91,87 @@ for (String s : result) {
+};
+TypeInformation[] typeInformations = new TypeInformation[] {
+	Types.LONG,
+	Types.FLOAT,
```
```suggestion
	Types.LONG,
```

qiaojialin commented on flink-tsfile-connector/README.md:
```diff
@@ -91,3 +91,87 @@ for (String s : result) {
+	Types.FLOAT,
+	Types.LONG,
+	Types.LONG,
+	Types.FLOAT,
```
```suggestion
	Types.LONG,
```

WeiZhong94 commented:
@sunjincheng121 @qiaojialin Thanks for your reply! I have correct the README.md. It seems that the travis test failure is not involved in this PR, so I just rebased this PR to trigger the test, hope this can work.
